### ISSUE TITLE
deep-diff does not handle cycles

### ISSUE THREAD

**Issue by 0x24a537r9** [flitbit/diff#62]
Great library! Turns out this is (almost) exactly the kind of thing I was looking for to build an optimized undo framework for my app. Sadly, I won't be able to use it since it appears deep-diff does not handle cycles:

See https://tonicdev.com/0x24a537r9/deep-diff-test for an example.

Clearly some sort of cycle _detection_ is implemented, otherwise this would trigger an infinite loop. But I see no reason why diffs could not be provided in this case for the non-cyclical level.

**Comment by kownacki** [comment]
Hi, @0x24a537r9

I don't agree. In this case those objects should be considered equal. Also using `_.isEqual` gives positive result. Your case can be simplified to this:

```
var a = {};
var b = {};
a.circ = b;
b.circ = a;

console.log(deepDiff.diff(a, b));
console.log(_.isEqual(a, b));
```

**Comment by 0x24a537r9** [comment]
Hmm, after some thought I suppose in the case I outlined, you're right that as long as deep-diff works when `a` and `b` have other differing properties (which it does), then they are functionally equivalent, insofar as their reference graphs are isomorphic.

On second thought, though, I've found a valid counter-example where the reference graph itself between two objects is *not* isomorphic, yet deep-diff (and Underscore) identifies the two objects as equivalent:

```
var deep = require('deep-diff')
var _ = require('underscore')

var a = {};
var b = {};
a.x = a;
a.y = b;
b.x = b;
b.y = b;

console.log('Deep diff: ' + deep.diff(a, b));
console.log('Underscore: ' + _.isEqual(a, b));
```

The reference graph for `a` and `b` can be found [here](http://i.imgur.com/jvJQpb0.png).

Or if having them be subgraphs is confusing, consider also:

```
var deep = require('deep-diff')
var _ = require('underscore')

var a = {};
var b = {};
a.l = b;
a.l = a;

var c = {};
c.l = c;

console.log('Deep diff: ' + deep.diff(a, c));
console.log('Underscore: ' + _.isEqual(a, c));
```

In both cases, both libraries identify nonisomorphic graphs as equivalent. In Underscore.js's annotated documentation for `_.equal()` [here](http://underscorejs.org/docs/underscore.html#section-125) they explicitly note:

> Assume equality for cyclic structures.

Given that the [graph isomorphism problem](https://en.wikipedia.org/wiki/Graph_isomorphism_problem) is difficult and the likelihood of needing this particular behavior is low, I think it's probably fine to go with Underscore's simplifying assumption that cycles are equivalent, even if technically unsafe (perhaps with a caveat in the documentation for these kinds of edge cases).

I'm fine with marking this as "working as intended" if you are though :)

**Comment by kownacki** [comment]
1: deep-diff doesn't handle all kinds of cycles (https://github.com/flitbit/diff/issues/74) so I wouldn't really make it as "working as intended".

2: Graph isomorphism problem is not applicable here because objects in javascript have named edges (properties) and we have starting points of both "graphs". So we never have to find which descendant of given node has to be checked.

3. Great observation of yours, that objects said by `_.isEqual` to be equivalent don't necessarily need to have their graph structure isomorphic! We can see it in your example which features both circularity and intersection. Even though this might be could be counter intuitive but actually it makes much sense in JS world. Just to simplify your first case: 

```javascript
var deep = require('deep-diff');
var _ = require('underscore');

var a = {};
var b = {};
a.x = b;
b.x = b;

console.log('Deep diff: ' + deep.diff(a, b));
console.log('Underscore: ' + _.isEqual(a, b));
```

`a` and `b` are different in a sense that `a` has access to `b` but `b` doesn't. But in "JS terms" **`a` and `b` both have single property `x` which references the same object** - clearly, they should be considered equal! It only just happens that under `x` we have `b` :).

4. Underscore/lodash isEqual seems to be the ultimate deep comparator for JS values. But I still can't find a fully sufficient diff generator which will work like isEqual but just showing what's different in objects. I don't even know all caveats although I've found a lot already (like in point 3).  Deep-diff could be the one but has problems with classes and more sophisticated circular structures, unfortunately.

**Comment by 0x24a537r9** [comment]
(To clarify, the reason I'm bringing all this up is because I actually wanted to use deep-diff for a graph theory application, so the 'shape' of the data structure is actually very important)

2: Good point--simpler than the graph isomorphism problem, but still probably difficult without the ability to get the actual reference id for an object (excluding nonstandard debugging APIs).

3: I also stumbled upon that example in thinking about this, but thought it would be distracting because of the identity issue you mentioned, so didn't bring it up. If you accept that `a` and `b` should be considered equal because their references are equal, consider how deep-diff also treats the following:

```
var a = {};
var b = {};
a.x = b;
b.x = b;
deep.diff(a, b);  // True

a.x = a;  // Change to a
// No change to b
deep.diff(a, b);  // Still true...
```

I find that disturbing since it means that changing a property of only one of two equal inputs (an inherently asymmetric operation) does not invalidate their equality. Now you may say something to the effect of "well, their infinite serializations are still equal", which is true. If "having equal infinite serializations" is your definition of equality, then changing `a.x` to an "equivalent" reference is effectively the same as multiplying a number by 1, or adding 0. But consider that that definition of equality may be too weak in some cases...

I would argue that for a stronger definition of equality, given any two mutually exclusive data structures `a` and `b`, for any pure function `f`, `a = b → f(a) = f(b)`. Said alternatively, if two mutually exclusive data structures are "equivalent", then we should also expect the results of applying a pure function to both to be equivalent. The contrapositive of this is that two mutually exclusive data structures should not be considered equal if the same pure function applied to both yields different results. Consider then the contradiction of:

```
var deep = require('deep-diff')

var a1 = {v: 1}, a2 = {v: 2}; a1.l = a2; a2.l = a2;
var b1 = {v: 1}, b2 = {v: 2}; b1.l = b2; b2.l = b1;

function f(x) {
    var sum = 0;
    sum += x.v; x = x.l;
    sum += x.v; x = x.l;
    sum += x.v; x = x.l;
    return sum;
}

console.log('Deep diff of a1, b1: ', deep.diff(a1, b1));
console.log('Deep diff of f(a1), f(b1): ', deep.diff(f(a1), f(b1)));
```

`a` and `b` are mutually exclusive, and `f()` is a pure function, yet `a = b` and `f(a) != f(b)`. This is why I think the 'shape' of the reference graph is so important (e.g. in the case of graph traversals), and why "having equal infinite serializations" is not sufficient for "strong equivalence". Consider if you were using deep-diff to implement undo/redo functions in a graph theory application (as I was planning to); sometimes undo/redo would simply fail, since deep-diff didn't notice any change between snapshots.

To be clear, it would take a lot of work to build an efficient algorithm to distinguish such structures in generality, especially in a language that allows reference equality testing but does not expose object reference ids themselves, so I think it's well within your rights as the library owner to say that that strength of equivalence is outside the scope of this project (as Underscore.js has). "Weak equivalence" is probably perfectly suitable for vast majority of applications (although a disclaimer might be useful).

4: Perhaps, but unfortunately it also suffers the same weakness highlighted above. I certainly think deep-diff is the best diff generator I've seen out there, though :)

**Comment by flitbit** [comment]
Thanks for the discussion @0x24a537r9 and @kownacki ; you are both too smart, too academic, or work on much harder problems than I can follow... I appreciate the brain-time you put in on this problem/issue, and since _brain-time is a developer's real stock-in-trade_, thank you. 

I recently fixed #74, referenced in the discussion, and added [examples/issue-62.js](https://github.com/flitbit/diff/blob/0219a5889b5ef027e46102328810aa59394c9f5a/examples/issue-62.js) showing that the [disturbing observations referred to by @0x24a537r9 (3)](https://github.com/flitbit/diff/issues/62#issuecomment-229549984) no longer hold.

I tried to be explicit that my interest when authoring this library was a reliable mechanism for detecting structural differences between objects. Clearly, such is complicated by cycles in the graph, and I took a simplistic approach to cycles.

I'll also note that I'm not going for _strong equivalence_ since javascript objects at rest or in transit are (usually) represented as JSON, and often revived and processed in languages that differ from the origin (in my original use case it was _javascript_ on one end and _erlang_ on the other). Nonetheless, #74 was a long-standing bug I should have cared for earlier.

I'm closing this issue. If there are lingering issues among the many ideas referenced, lets illustrate the remainder, if they exist, in new issues. ~Thanks.

**Comment by 0x24a537r9** [comment]
Haha, I actually totally agree with your approach--I think it is more than sufficient for the vast majority of cases, which is the right way to prioritize features in this kind of library. The amount of code needed to support this kind of topological equality comparison performantly is, I think, rightly outside of the scope of this project. Thanks for fixing #74 though!
